{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/M278428/opt/anaconda3/envs/test_dev_vrs_installation/lib/python3.9/site-packages/python_jsonschema_objects/__init__.py:49: UserWarning: Schema version http://json-schema.org/draft-07/schema not recognized. Some keywords and features may not be supported.\n",
      "  warnings.warn(\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "from src.extra_translate import extra\n",
    "from src.variation_services_rest_db import VariationServicesRESTDataProxy\n",
    "convert = extra()\n",
    "converting_hgvs = VariationServicesRESTDataProxy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "from ga4gh.vrs.dataproxy import SeqRepoRESTDataProxy\n",
    "from ga4gh.vrs.extras.translator import Translator\n",
    "from ga4gh.vrs.extras.variation_normalizer_rest_dp import VariationNormalizerRESTDataProxy\n",
    "vnorm = VariationNormalizerRESTDataProxy()\n",
    "\n",
    "seqrepo_rest_service_url = \"https://services.genomicmedlab.org/seqrepo\"\n",
    "dp = SeqRepoRESTDataProxy(base_url=seqrepo_rest_service_url)\n",
    "tlr = Translator(dp)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Variant Region ID</th>\n",
       "      <th>Type</th>\n",
       "      <th>Clinical Assertion</th>\n",
       "      <th>Location</th>\n",
       "      <th>Genes in region</th>\n",
       "      <th>description</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>nsv3875336</td>\n",
       "      <td>copy number variation</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>GRCh37 chr1: 143,721,526-149,232,481 , GRCh38....</td>\n",
       "      <td>LOC101060796</td>\n",
       "      <td>1q21.1-21.2(chr1:143721526-149232481)x</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>nsv6312995</td>\n",
       "      <td>copy number variation</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>GRCh37 chrX: 31,645,770-31,792,329 , GRCh38.p1...</td>\n",
       "      <td>DMD</td>\n",
       "      <td>NC_000023.10:g.(?_31645770)_(31792329_?)del</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>nsv5672872</td>\n",
       "      <td>copy number variation</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>GRCh37 chr17: 29,422,314-29,548,957 , GRCh38.p...</td>\n",
       "      <td>NF1</td>\n",
       "      <td>NC_000017.10:g.(?_29422314)_(29548957_?)del</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>nsv6137852</td>\n",
       "      <td>copy number variation</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>GRCh38 chr16: 3,727,580-3,851,110 , GRCh37.p13...</td>\n",
       "      <td>CREBBP</td>\n",
       "      <td>16p13.3(chr16:3727580-3851110)x1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>nsv7093597</td>\n",
       "      <td>insertion</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>GRCh37 chr17: 29,550,464-29,550,464 , GRCh38 c...</td>\n",
       "      <td>NF1</td>\n",
       "      <td>NM_001042492.3(NF1):c.1738_1739insTTTTTTTTTTTT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>nsv6314514</td>\n",
       "      <td>complex chromosomal rearrangement</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>GRCh37 chr7: 128,114,028-128,114,028 , GRCh37 ...</td>\n",
       "      <td>ZNF507</td>\n",
       "      <td>46;XY;t(7;19)(q32;q13.1)dn</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>nsv3919275</td>\n",
       "      <td>insertion</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>GRCh37 chr7: 42,005,900-42,005,900 , GRCh38 ch...</td>\n",
       "      <td>GLI3</td>\n",
       "      <td>NM_000168.5(GLI3):c.2770_2771insNC_012920.1:g....</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>nsv4684390</td>\n",
       "      <td>delins</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>GRCh37 chr9: 108,484,281-108,505,262 , GRCh38 ...</td>\n",
       "      <td>TMEM38B</td>\n",
       "      <td>NC_000009.11:g.108484281_108505262delinsAATTAA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>nsv4578697</td>\n",
       "      <td>delins</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>GRCh37 chr16: 1,569,145-1,571,812 , GRCh38 chr...</td>\n",
       "      <td>IFT140</td>\n",
       "      <td>NC_000016.9:g.1569145_1571812delinsG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>nsv4436066</td>\n",
       "      <td>complex substitution</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>GRCh38.p12 chr9: 78,228,782-78,244,763 , GRCh3...</td>\n",
       "      <td>CEP78</td>\n",
       "      <td>46,XX,der(9)(q21.2,q21.2).seq[GRCh37/hg19]der(...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>nsv6638079</td>\n",
       "      <td>inversion</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>GRCh37 chr5: 112,154,360-112,155,231 , GRCh38 ...</td>\n",
       "      <td>APC</td>\n",
       "      <td>NC_000005.10:g.112818663_112819534inv872</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>nsv6314631</td>\n",
       "      <td>complex chromosomal rearrangement</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>GRCh38.p12 chr12: 13,802,882-13,802,882 , GRCh...</td>\n",
       "      <td>GRIN2B</td>\n",
       "      <td>46;XY;inv(12)(p12.1q21.2)dn</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>nsv6314282</td>\n",
       "      <td>complex chromosomal rearrangement</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>GRCh37 chr11: 46,619,320-46,619,320 , GRCh37 c...</td>\n",
       "      <td>EHMT1</td>\n",
       "      <td>46;XY;t(9;11)(q34;p11.2)dn</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>nsv7093439</td>\n",
       "      <td>delins</td>\n",
       "      <td>Likely pathogenic</td>\n",
       "      <td>GRCh38 chr10: 101,017,287-101,018,201 , GRCh37...</td>\n",
       "      <td>PDZD7</td>\n",
       "      <td>NC_000010.11:g.101017287_101018201delinsT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>nsv7093572</td>\n",
       "      <td>inversion</td>\n",
       "      <td>Likely pathogenic</td>\n",
       "      <td>GRCh37 chr20: 49,515,761-49,525,309 , GRCh38 c...</td>\n",
       "      <td>ADNP</td>\n",
       "      <td>NC_000020.11:g.50899224_50908772inv9549</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>nsv4458037</td>\n",
       "      <td>tandem duplication</td>\n",
       "      <td>Uncertain significance</td>\n",
       "      <td>GRCh37 chr9: 135,803,187-135,807,265 , GRCh38....</td>\n",
       "      <td>TSC1</td>\n",
       "      <td>NC_000009.11:g.(135803187_135803191)_(13580726...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>nsv6314259</td>\n",
       "      <td>insertion</td>\n",
       "      <td>Uncertain significance</td>\n",
       "      <td>GRCh37 chr7: 130,023,531-130,023,531 , GRCh38 ...</td>\n",
       "      <td>CPA1</td>\n",
       "      <td>NM_001868.4(CPA1):c.607_608insGCCGGGCGCGGTGGCT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>nsv6634698</td>\n",
       "      <td>delins</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>GRCh37 chr3: 37,090,048-37,090,048 , GRCh38 ch...</td>\n",
       "      <td>MLH1</td>\n",
       "      <td>NM_000249.4(MLH1):c.1937delinsTTGGAGGGACTGCCCC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>nsv6314306</td>\n",
       "      <td>complex chromosomal rearrangement</td>\n",
       "      <td>Likely pathogenic</td>\n",
       "      <td>GRCh38.p12 chr4: 11,160,052-11,160,052 , GRCh3...</td>\n",
       "      <td>PBX1</td>\n",
       "      <td>46;XY;t(1;4)(q21.3;p15.2)dn</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>nsv6314613</td>\n",
       "      <td>inversion</td>\n",
       "      <td>Uncertain significance</td>\n",
       "      <td>GRCh37 chrX: 13,786,339-13,786,405 , GRCh38 ch...</td>\n",
       "      <td>OFD1</td>\n",
       "      <td>NC_000023.11:g.13768220_13768286inv67</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>nsv6314434</td>\n",
       "      <td>complex chromosomal rearrangement</td>\n",
       "      <td>Uncertain significance</td>\n",
       "      <td>GRCh38.p12 chr14: 64,254,366-64,254,366 , GRCh...</td>\n",
       "      <td>ESR2</td>\n",
       "      <td>46;XX;inv(14)(q24.1q32.1)dn</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>esv3302965</td>\n",
       "      <td>tandem duplication</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NCBI36 chr17: 35,950,953-35,962,203 , GRCh37.p...</td>\n",
       "      <td>CCR7</td>\n",
       "      <td>NC_000017.10:g.(38697428_38697628)_(38708477_3...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>esv3302998</td>\n",
       "      <td>tandem duplication</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NCBI36 chr22: 40,948,203-40,949,546 , GRCh37.p...</td>\n",
       "      <td>TCF20</td>\n",
       "      <td>NC_000022.11:g.(42222254_42222454)_(42223396_4...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   Variant Region ID                               Type  \\\n",
       "0         nsv3875336              copy number variation   \n",
       "1         nsv6312995              copy number variation   \n",
       "2         nsv5672872              copy number variation   \n",
       "3         nsv6137852              copy number variation   \n",
       "4         nsv7093597                          insertion   \n",
       "5         nsv6314514  complex chromosomal rearrangement   \n",
       "6         nsv3919275                          insertion   \n",
       "7         nsv4684390                             delins   \n",
       "8         nsv4578697                             delins   \n",
       "9         nsv4436066               complex substitution   \n",
       "10        nsv6638079                          inversion   \n",
       "11        nsv6314631  complex chromosomal rearrangement   \n",
       "12        nsv6314282  complex chromosomal rearrangement   \n",
       "13        nsv7093439                             delins   \n",
       "14        nsv7093572                          inversion   \n",
       "15        nsv4458037                 tandem duplication   \n",
       "16        nsv6314259                          insertion   \n",
       "17        nsv6634698                             delins   \n",
       "18        nsv6314306  complex chromosomal rearrangement   \n",
       "19        nsv6314613                          inversion   \n",
       "20        nsv6314434  complex chromosomal rearrangement   \n",
       "21        esv3302965                 tandem duplication   \n",
       "22        esv3302998                 tandem duplication   \n",
       "\n",
       "        Clinical Assertion                                           Location  \\\n",
       "0               Pathogenic  GRCh37 chr1: 143,721,526-149,232,481 , GRCh38....   \n",
       "1               Pathogenic  GRCh37 chrX: 31,645,770-31,792,329 , GRCh38.p1...   \n",
       "2               Pathogenic  GRCh37 chr17: 29,422,314-29,548,957 , GRCh38.p...   \n",
       "3               Pathogenic  GRCh38 chr16: 3,727,580-3,851,110 , GRCh37.p13...   \n",
       "4               Pathogenic  GRCh37 chr17: 29,550,464-29,550,464 , GRCh38 c...   \n",
       "5               Pathogenic  GRCh37 chr7: 128,114,028-128,114,028 , GRCh37 ...   \n",
       "6               Pathogenic  GRCh37 chr7: 42,005,900-42,005,900 , GRCh38 ch...   \n",
       "7               Pathogenic  GRCh37 chr9: 108,484,281-108,505,262 , GRCh38 ...   \n",
       "8               Pathogenic  GRCh37 chr16: 1,569,145-1,571,812 , GRCh38 chr...   \n",
       "9               Pathogenic  GRCh38.p12 chr9: 78,228,782-78,244,763 , GRCh3...   \n",
       "10              Pathogenic  GRCh37 chr5: 112,154,360-112,155,231 , GRCh38 ...   \n",
       "11              Pathogenic  GRCh38.p12 chr12: 13,802,882-13,802,882 , GRCh...   \n",
       "12              Pathogenic  GRCh37 chr11: 46,619,320-46,619,320 , GRCh37 c...   \n",
       "13       Likely pathogenic  GRCh38 chr10: 101,017,287-101,018,201 , GRCh37...   \n",
       "14       Likely pathogenic  GRCh37 chr20: 49,515,761-49,525,309 , GRCh38 c...   \n",
       "15  Uncertain significance  GRCh37 chr9: 135,803,187-135,807,265 , GRCh38....   \n",
       "16  Uncertain significance  GRCh37 chr7: 130,023,531-130,023,531 , GRCh38 ...   \n",
       "17              Pathogenic  GRCh37 chr3: 37,090,048-37,090,048 , GRCh38 ch...   \n",
       "18       Likely pathogenic  GRCh38.p12 chr4: 11,160,052-11,160,052 , GRCh3...   \n",
       "19  Uncertain significance  GRCh37 chrX: 13,786,339-13,786,405 , GRCh38 ch...   \n",
       "20  Uncertain significance  GRCh38.p12 chr14: 64,254,366-64,254,366 , GRCh...   \n",
       "21                     NaN  NCBI36 chr17: 35,950,953-35,962,203 , GRCh37.p...   \n",
       "22                     NaN  NCBI36 chr22: 40,948,203-40,949,546 , GRCh37.p...   \n",
       "\n",
       "   Genes in region                                        description  \n",
       "0     LOC101060796             1q21.1-21.2(chr1:143721526-149232481)x  \n",
       "1              DMD        NC_000023.10:g.(?_31645770)_(31792329_?)del  \n",
       "2              NF1        NC_000017.10:g.(?_29422314)_(29548957_?)del  \n",
       "3           CREBBP                   16p13.3(chr16:3727580-3851110)x1  \n",
       "4              NF1  NM_001042492.3(NF1):c.1738_1739insTTTTTTTTTTTT...  \n",
       "5           ZNF507                         46;XY;t(7;19)(q32;q13.1)dn  \n",
       "6             GLI3  NM_000168.5(GLI3):c.2770_2771insNC_012920.1:g....  \n",
       "7          TMEM38B  NC_000009.11:g.108484281_108505262delinsAATTAA...  \n",
       "8           IFT140               NC_000016.9:g.1569145_1571812delinsG  \n",
       "9            CEP78  46,XX,der(9)(q21.2,q21.2).seq[GRCh37/hg19]der(...  \n",
       "10             APC           NC_000005.10:g.112818663_112819534inv872  \n",
       "11          GRIN2B                        46;XY;inv(12)(p12.1q21.2)dn  \n",
       "12           EHMT1                         46;XY;t(9;11)(q34;p11.2)dn  \n",
       "13           PDZD7          NC_000010.11:g.101017287_101018201delinsT  \n",
       "14            ADNP            NC_000020.11:g.50899224_50908772inv9549  \n",
       "15            TSC1  NC_000009.11:g.(135803187_135803191)_(13580726...  \n",
       "16            CPA1  NM_001868.4(CPA1):c.607_608insGCCGGGCGCGGTGGCT...  \n",
       "17            MLH1  NM_000249.4(MLH1):c.1937delinsTTGGAGGGACTGCCCC...  \n",
       "18            PBX1                        46;XY;t(1;4)(q21.3;p15.2)dn  \n",
       "19            OFD1              NC_000023.11:g.13768220_13768286inv67  \n",
       "20            ESR2                        46;XX;inv(14)(q24.1q32.1)dn  \n",
       "21            CCR7  NC_000017.10:g.(38697428_38697628)_(38708477_3...  \n",
       "22           TCF20  NC_000022.11:g.(42222254_42222454)_(42223396_4...  "
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = pd.read_csv('/Users/M278428/Documents/educational_notebooks/data/dbVar_structural_variants.csv',dtype = str)\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "def from_hgvs_to_spdi(expression):\n",
    "    try: \n",
    "        # Converting a allele in SPDI syntax to the right-shifted HGVS notation\n",
    "        return converting_hgvs.hgvs_to_spdi(expression)\n",
    "    except Exception as e: \n",
    "        # returns error produce by NCBI API \n",
    "        return 'Error in expression {}'.format(e)     "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "convert_to_spdi = []\n",
    "correct_spdi_expression= []\n",
    "spdi_expressions = []\n",
    "\n",
    "index_of_incorrect_spdi_expr= []\n",
    "index_of_correct_spdi_expr= []\n",
    "\n",
    "for sample in df['description']:\n",
    "    convert_to_spdi.append(from_hgvs_to_spdi(sample)) \n",
    "\n",
    "for idx,spdi in enumerate(convert_to_spdi):\n",
    "    index_of_incorrect_spdi_expr.append(idx)\n",
    "    if isinstance(spdi,list):\n",
    "        index_of_correct_spdi_expr.append(idx)\n",
    "        correct_spdi_expression.append(spdi)\n",
    "\n",
    "for sublist in correct_spdi_expression:\n",
    "    spdi_expressions.extend(sublist)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "spdi_expressions.append('hahahahha')\n",
    "spdi_expressions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "vrs_alleles = []\n",
    "hgvs_expressions = []\n",
    "\n",
    "for expression in spdi_expressions:\n",
    "    try:\n",
    "        trans = tlr.translate_from(expression, 'spdi')\n",
    "        vrs_alleles.append(trans)\n",
    "    except Exception as e:\n",
    "        # returns error produce by translate_from method\n",
    "        vrs_alleles.append('Error in expression {}'.format(e))\n",
    "\n",
    "# Removing string values \n",
    "vrs_alleles = [x for x in vrs_alleles if not isinstance(x, str)]\n",
    "\n",
    "hgvs_expressions = []\n",
    "for allele in vrs_alleles:  \n",
    "    try:  \n",
    "        # translating the vrs objects to hgvs expressions\n",
    "        hgvs_expressions.append(vnorm.to_hgvs(allele, 'refseq'))\n",
    "    except Exception as e:\n",
    "        # returns error produce by to_hgvs method\n",
    "        hgvs_expressions.append('Error in expression {}'.format(e))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[['NC_000009.11:g.108484281_108505262delinsAATTAAGGTATA'],\n",
       " ['NC_000016.9:g.1569145_1571812delinsG'],\n",
       " ['NC_000005.10:g.112818663_112819534inv'],\n",
       " ['NC_000010.11:g.101017287_101018201delinsT'],\n",
       " 'Error in expression Variation normalizer returned the status code: 403.',\n",
       " ['NC_000023.11:g.13768220_13768286inv']]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "hgvs_expressions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>input_description</th>\n",
       "      <th>ga4gh_id</th>\n",
       "      <th>vrs_allele_object</th>\n",
       "      <th>spdi_expression</th>\n",
       "      <th>fullynorm_hgvs</th>\n",
       "      <th>rightshift_hgvs</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>NC_000009.11:g.108484281_108505262delinsAATTAA...</td>\n",
       "      <td>(ga4gh:VA.NEdt1t2Ak_hxAbdsgrrchZKbWFE05SkC)</td>\n",
       "      <td>{'_id': 'ga4gh:VA.NEdt1t2Ak_hxAbdsgrrchZKbWFE0...</td>\n",
       "      <td>[NC_000009.11:108484280:TAGTGCCTCTGTAAATAAAATA...</td>\n",
       "      <td>[NC_000009.11:g.108484281_108505262delinsAATTA...</td>\n",
       "      <td>Error in expression Variation Services returne...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>NC_000016.9:g.1569145_1571812delinsG</td>\n",
       "      <td>(ga4gh:VA.61NVIrpqJfF0bd80g-3-3hWO10IGpOqb)</td>\n",
       "      <td>{'_id': 'ga4gh:VA.61NVIrpqJfF0bd80g-3-3hWO10IG...</td>\n",
       "      <td>[NC_000016.9:1569144:ACCCCAGAATCGACACGTGCAGGTG...</td>\n",
       "      <td>[NC_000016.9:g.1569145_1571812delinsG]</td>\n",
       "      <td>NC_000016.9:g.1569145_1571812delinsG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>NC_000005.10:g.112818663_112819534inv872</td>\n",
       "      <td>(ga4gh:VA.lRyFtt5JyBcrfTcQkvrch32ChQYdCF06)</td>\n",
       "      <td>{'_id': 'ga4gh:VA.lRyFtt5JyBcrfTcQkvrch32ChQYd...</td>\n",
       "      <td>[NC_000005.10:112818662:TCTAGGCAAACAGCACTAACAG...</td>\n",
       "      <td>[NC_000005.10:g.112818663_112819534inv]</td>\n",
       "      <td>NC_000005.10:g.112818663_112819534inv</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>NC_000010.11:g.101017287_101018201delinsT</td>\n",
       "      <td>(ga4gh:VA.KqqgYuCedDo9_DMeGsbm8AYBGyUdxs2u)</td>\n",
       "      <td>{'_id': 'ga4gh:VA.KqqgYuCedDo9_DMeGsbm8AYBGyUd...</td>\n",
       "      <td>[NC_000010.11:101017286:AACTCATGGGAGAGATGGAGGG...</td>\n",
       "      <td>[NC_000010.11:g.101017287_101018201delinsT]</td>\n",
       "      <td>NC_000010.11:g.101017287_101018201delinsT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>NC_000020.11:g.50899224_50908772inv9549</td>\n",
       "      <td>(ga4gh:VA.VaSFeT-EB60a3fqRqzeRWhOzZZmcETBe)</td>\n",
       "      <td>{'_id': 'ga4gh:VA.VaSFeT-EB60a3fqRqzeRWhOzZZmc...</td>\n",
       "      <td>[NC_000020.11:50899223:CTTGCTTTTGTCCCCCAGGCTGG...</td>\n",
       "      <td>Error in expression Variation normalizer retur...</td>\n",
       "      <td>Error in expression Variation Services returne...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>NC_000023.11:g.13768220_13768286inv67</td>\n",
       "      <td>(ga4gh:VA.xeZrD37Ki00nCEcGA-mr7XHMMyuGmSiS)</td>\n",
       "      <td>{'_id': 'ga4gh:VA.xeZrD37Ki00nCEcGA-mr7XHMMyuG...</td>\n",
       "      <td>[NC_000023.11:13768219:ATAAGGTGCCAGTGCCATGGGCA...</td>\n",
       "      <td>[NC_000023.11:g.13768220_13768286inv]</td>\n",
       "      <td>NC_000023.11:g.13768220_13768286inv</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                    input_description  \\\n",
       "7   NC_000009.11:g.108484281_108505262delinsAATTAA...   \n",
       "8                NC_000016.9:g.1569145_1571812delinsG   \n",
       "10           NC_000005.10:g.112818663_112819534inv872   \n",
       "13          NC_000010.11:g.101017287_101018201delinsT   \n",
       "14            NC_000020.11:g.50899224_50908772inv9549   \n",
       "19              NC_000023.11:g.13768220_13768286inv67   \n",
       "\n",
       "                                       ga4gh_id  \\\n",
       "7   (ga4gh:VA.NEdt1t2Ak_hxAbdsgrrchZKbWFE05SkC)   \n",
       "8   (ga4gh:VA.61NVIrpqJfF0bd80g-3-3hWO10IGpOqb)   \n",
       "10  (ga4gh:VA.lRyFtt5JyBcrfTcQkvrch32ChQYdCF06)   \n",
       "13  (ga4gh:VA.KqqgYuCedDo9_DMeGsbm8AYBGyUdxs2u)   \n",
       "14  (ga4gh:VA.VaSFeT-EB60a3fqRqzeRWhOzZZmcETBe)   \n",
       "19  (ga4gh:VA.xeZrD37Ki00nCEcGA-mr7XHMMyuGmSiS)   \n",
       "\n",
       "                                    vrs_allele_object  \\\n",
       "7   {'_id': 'ga4gh:VA.NEdt1t2Ak_hxAbdsgrrchZKbWFE0...   \n",
       "8   {'_id': 'ga4gh:VA.61NVIrpqJfF0bd80g-3-3hWO10IG...   \n",
       "10  {'_id': 'ga4gh:VA.lRyFtt5JyBcrfTcQkvrch32ChQYd...   \n",
       "13  {'_id': 'ga4gh:VA.KqqgYuCedDo9_DMeGsbm8AYBGyUd...   \n",
       "14  {'_id': 'ga4gh:VA.VaSFeT-EB60a3fqRqzeRWhOzZZmc...   \n",
       "19  {'_id': 'ga4gh:VA.xeZrD37Ki00nCEcGA-mr7XHMMyuG...   \n",
       "\n",
       "                                      spdi_expression  \\\n",
       "7   [NC_000009.11:108484280:TAGTGCCTCTGTAAATAAAATA...   \n",
       "8   [NC_000016.9:1569144:ACCCCAGAATCGACACGTGCAGGTG...   \n",
       "10  [NC_000005.10:112818662:TCTAGGCAAACAGCACTAACAG...   \n",
       "13  [NC_000010.11:101017286:AACTCATGGGAGAGATGGAGGG...   \n",
       "14  [NC_000020.11:50899223:CTTGCTTTTGTCCCCCAGGCTGG...   \n",
       "19  [NC_000023.11:13768219:ATAAGGTGCCAGTGCCATGGGCA...   \n",
       "\n",
       "                                       fullynorm_hgvs  \\\n",
       "7   [NC_000009.11:g.108484281_108505262delinsAATTA...   \n",
       "8              [NC_000016.9:g.1569145_1571812delinsG]   \n",
       "10            [NC_000005.10:g.112818663_112819534inv]   \n",
       "13        [NC_000010.11:g.101017287_101018201delinsT]   \n",
       "14  Error in expression Variation normalizer retur...   \n",
       "19              [NC_000023.11:g.13768220_13768286inv]   \n",
       "\n",
       "                                      rightshift_hgvs  \n",
       "7   Error in expression Variation Services returne...  \n",
       "8                NC_000016.9:g.1569145_1571812delinsG  \n",
       "10              NC_000005.10:g.112818663_112819534inv  \n",
       "13          NC_000010.11:g.101017287_101018201delinsT  \n",
       "14  Error in expression Variation Services returne...  \n",
       "19                NC_000023.11:g.13768220_13768286inv  "
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "allForms = {\n",
    "    'input_description' : df.iloc[index_of_correct_spdi_expr]['description'],\n",
    "    'ga4gh_id' : [],\n",
    "    'vrs_allele_object' : [],\n",
    "    'spdi_expression' : correct_spdi_expression,\n",
    "    'fullynorm_hgvs' : hgvs_expressions,\n",
    "    'rightshift_hgvs' : []\n",
    "    }\n",
    "\n",
    "vrs_object_list = []\n",
    "\n",
    "for sample in spdi_expressions:\n",
    "    allForms['rightshift_hgvs'].append(convert.to_rightshift_hgvs(sample))\n",
    "    # allForms['fullynorm_hgvs'].append(convert.to_fullynorm_hgvs(sample))\n",
    "    vrs_object_list.append(convert.to_vrs_allele(sample))  \n",
    "    \n",
    "for vrs_objects in vrs_object_list:\n",
    "    for vrs_id in vrs_objects.keys():\n",
    "        allForms['ga4gh_id'].append(vrs_objects.keys())\n",
    "        allForms['vrs_allele_object'].append(vrs_objects[vrs_id])\n",
    "\n",
    "\n",
    "allFormDF = pd.DataFrame.from_dict(allForms)\n",
    "allFormDF"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "7     NC_000009.11:g.108484281_108505262delinsAATTAA...\n",
       "8                  NC_000016.9:g.1569145_1571812delinsG\n",
       "10             NC_000005.10:g.112818663_112819534inv872\n",
       "13            NC_000010.11:g.101017287_101018201delinsT\n",
       "14              NC_000020.11:g.50899224_50908772inv9549\n",
       "19                NC_000023.11:g.13768220_13768286inv67\n",
       "Name: description, dtype: object"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    " df.iloc[index_of_correct_spdi_expr]['description']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0                1q21.1-21.2(chr1:143721526-149232481)x\n",
       "1           NC_000023.10:g.(?_31645770)_(31792329_?)del\n",
       "2           NC_000017.10:g.(?_29422314)_(29548957_?)del\n",
       "3                      16p13.3(chr16:3727580-3851110)x1\n",
       "4     NM_001042492.3(NF1):c.1738_1739insTTTTTTTTTTTT...\n",
       "5                            46;XY;t(7;19)(q32;q13.1)dn\n",
       "6     NM_000168.5(GLI3):c.2770_2771insNC_012920.1:g....\n",
       "7     NC_000009.11:g.108484281_108505262delinsAATTAA...\n",
       "8                  NC_000016.9:g.1569145_1571812delinsG\n",
       "9     46,XX,der(9)(q21.2,q21.2).seq[GRCh37/hg19]der(...\n",
       "10             NC_000005.10:g.112818663_112819534inv872\n",
       "11                          46;XY;inv(12)(p12.1q21.2)dn\n",
       "12                           46;XY;t(9;11)(q34;p11.2)dn\n",
       "13            NC_000010.11:g.101017287_101018201delinsT\n",
       "14              NC_000020.11:g.50899224_50908772inv9549\n",
       "15    NC_000009.11:g.(135803187_135803191)_(13580726...\n",
       "16    NM_001868.4(CPA1):c.607_608insGCCGGGCGCGGTGGCT...\n",
       "17    NM_000249.4(MLH1):c.1937delinsTTGGAGGGACTGCCCC...\n",
       "18                          46;XY;t(1;4)(q21.3;p15.2)dn\n",
       "19                NC_000023.11:g.13768220_13768286inv67\n",
       "20                          46;XX;inv(14)(q24.1q32.1)dn\n",
       "21    NC_000017.10:g.(38697428_38697628)_(38708477_3...\n",
       "22    NC_000022.11:g.(42222254_42222454)_(42223396_4...\n",
       "Name: description, dtype: object"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.iloc[index_of_incorrect_spdi_expr]['description']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "test_dev_vrs_installation",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.15"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
